File: Sources/ATTNUserIdentity.m
//
//  ATTNUserIdentifiers.h
//  Example
//
//  Created by Wyatt Davis on 11/9/22.
//

#import <Foundation/Foundation.h>

#import "ATTNParameterValidation.h"
#import "ATTNUserIdentity.h"
#import "ATTNVisitorService.h"
#import "ATTNConstants.h"

@implementation ATTNUserIdentity {
  ATTNVisitorService *_visitorService;
}

- (id)init {
  return [self initWithIdentifiers:@{}];
}

- (id)initWithIdentifiers:(nonnull NSDictionary *)identifiers {
  if (self = [super init]) {
    [self validateIdentifiers:identifiers];
    _identifiers = identifiers;
    _visitorService = [[ATTNVisitorService alloc] init];
    _visitorId = [_visitorService getVisitorId];
  }
  return self;
}

- (void)mergeIdentifiers:(nonnull NSDictionary *)newIdentifiers {
  [self validateIdentifiers:newIdentifiers];

  NSMutableDictionary *currentIdentifiersCopy = [_identifiers mutableCopy];
  [currentIdentifiersCopy addEntriesFromDictionary:newIdentifiers];
  _identifiers = currentIdentifiersCopy;
}

- (void)validateIdentifiers:(nonnull NSDictionary *)identifiers {
  NSArray *keys = @[ IDENTIFIER_TYPE_CLIENT_USER_ID,
                     IDENTIFIER_TYPE_PHONE,
                     IDENTIFIER_TYPE_EMAIL,
                     IDENTIFIER_TYPE_SHOPIFY_ID,
                     IDENTIFIER_TYPE_KLAVIYO_ID ];
  for (id key in keys) {
    [ATTNParameterValidation verifyStringOrNil:identifiers[key] inputName:key];
  }
  [ATTNParameterValidation verify1DStringDictionaryOrNil:identifiers[IDENTIFIER_TYPE_CUSTOM_IDENTIFIERS] inputName:IDENTIFIER_TYPE_CUSTOM_IDENTIFIERS];
}

- (void)clearUser {
  _identifiers = @{};
  _visitorId = [_visitorService createNewVisitorId];
}


@end

File: Sources/Public/ATTNUserIdentity.swift
//
//  ATTNUserIdentity.swift
//  attentive-ios-sdk-framework
//
//  Created by Vladimir - Work on 2024-05-29.
//

import Foundation

@objc(ATTNUserIdentity)
public final class ATTNUserIdentity: NSObject {
  @objc public var identifiers: [String: Any] {
    willSet {
      validate(identifiers: newValue)
    }
  }

  @objc public var visitorId: String
  private let visitorService: ATTNVisitorService

  @objc
  public override convenience init() {
    self.init(identifiers: [:])
  }

  @objc(initWithIdentifiers:)
  public init(identifiers: [String: Any]) {
    self.visitorService = .init()
    self.identifiers = identifiers
    self.visitorId = self.visitorService.getVisitorId()
    super.init()
  }

  @objc
  public func clearUser() {
    identifiers = [:]
    visitorId = visitorService.createNewVisitorId()
  }

  @objc
  public func mergeIdentifiers(_ newIdentifiers: [String: Any]) {
    validate(identifiers: newIdentifiers)

    var currentIdentifiers = identifiers
    // In case of a key conflict, the new value from newIdentifiers should be used.
    currentIdentifiers.merge(newIdentifiers) { (_, new) in new }
    identifiers = currentIdentifiers
  }
}

fileprivate extension ATTNUserIdentity {
  func validate(identifiers: [String: Any]) {
    ATTNIdentifierType.allKeys.forEach { currentKey in
      ATTNParameterValidation.verifyStringOrNil(
        identifiers[currentKey] as? NSObject,
        inputName: currentKey
      )
    }

    ATTNParameterValidation.verify1DStringDictionaryOrNil(
      identifiers[ATTNIdentifierType.customIdentifiers] as? NSDictionary,
      inputName: ATTNIdentifierType.customIdentifiers)
  }
}
